### PR TITLE
Vue3:fix reactive decorators

### LINKED ISSUE #21235: Addon-controls/Vue3: Control update while using decorators causes component to unmount/remount
### Describe the bug

When toggling a control value, the component unmounts and then remounts, preventing internal component state from being persisted. This prevents things like transitions and animations from being possible to represent via the controls.

### To Reproduce

https://github.com/treardon17/storybook-decorator-bug

### System

```shell
System:
    OS: macOS 13.2.1
    CPU: (12) arm64 Apple M2 Max
  Binaries:
    Node: 18.13.0 - ~/.nvm/versions/node/v18.13.0/bin/node
    Yarn: 1.22.19 - ~/.nvm/versions/node/v18.13.0/bin/yarn
    npm: 8.19.3 - ~/.nvm/versions/node/v18.13.0/bin/npm
  Browsers:
    Chrome: 110.0.5481.177
    Firefox: 108.0.1
    Safari: 16.3
  npmPackages:
    @storybook/addon-essentials: ^7.0.0-beta.53 => 7.0.0-beta.53
    @storybook/addon-interactions: ^7.0.0-beta.53 => 7.0.0-beta.53
    @storybook/addon-links: ^7.0.0-beta.53 => 7.0.0-beta.53
    @storybook/blocks: ^7.0.0-alpha.8 => 7.0.0-alpha.8
    @storybook/testing-library: ^0.0.14-next.1 => 0.0.14-next.1
    @storybook/vue3: ^7.0.0-beta.53 => 7.0.0-beta.53
    @storybook/vue3-vite: ^7.0.0-beta.53 => 7.0.0-beta.53
```

### Additional context

This issue is very similar to https://github.com/storybookjs/storybook/issues/13913, but a slight variation. Everything works fine when not using decorators (ie. the vue component mounts once and does not unmount when changing the args via the controls), but as soon as a decorator is added the components are now mounted/unmounted continuously when changing args via the controls.

Toggling this line on/off will demonstrate the behavior. https://github.com/treardon17/storybook-decorator-bug/blob/bd117642e0cb586a1d9d79966f76da22e83367de/.storybook/preview.ts#L28

Only mounting the component once allows for tests that watch props or deal with animations that require state, and using decorators allows for libraries like Vuetify to be able to inject a theme to the child components, so ideally both would work.

### PR DESCRIPTION
**Pull request by chakAs3**
Closes #21235 #21221 #21812 

## What I did

fixing losing reactivity when using decorators ( global decorator ex: vuetify , function story decorator , template story decorator )
now vue stories support 👍 :

- Reactive args  with default render function or custom render function, render function can be return (  Functional Component, Concrete Component , Options Component )
- Reactive args with story decorators, or global decorators, decorators are functions that return Functional Component, Concrete Component with template, or Options Component .
- Reactive named slots and scoped slots with default render or custom render function 

PR has a good tests coverage for these 3 features, but you can add some others if you feel they can be relevant 

## Checklist

- [ ] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [ ] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [x] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

### CODE REVIEW

File: code/renderers/vue3/template/stories/ReactiveDecorators.stories.ts [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        },
+        data() {
+          return { args: context.args };
+        },
+        template: '<div><h2>Decorator using label: {{args.label}}</h2><story/></div>',
+      });
+    },
+  ],
+};
+
```
Comment (chakAs3) [inline code/renderers/vue3/template/stories/ReactiveDecorators.stories.ts:86]: As i stated earlier decorators should be a Function that returns vnode  or a Component. 
Using VNode for component that wraps storyFn, you story will not be reactive. however the decorator is reactive in props level 
The secret for reactivity is to be Wrapped in Function.

File: code/renderers/vue3/template/stories/ReactiveDecorators.stories.ts [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+          story: storyFn(context.args),
+        },
+        props: ['label'],
+        template: '<div><div>Decorator using label: {{label}}</div><story/></div>',
+      });
+    },
+  ],
+};
+
+export const DecoratorFunctionalComponent: Story = {
```
Comment (chakAs3) [inline code/renderers/vue3/template/stories/ReactiveDecorators.stories.ts:101]: @kasperpeulen Here you want to test a functional component but you use the Input to check if state is preseved ?
Normally Functional component is pure function so there is no state or lifecycle it just props -> vnode so 100% your input loses his value 
However the component in reactive to changes and it is not remounted

File: code/renderers/vue3/template/stories/ReactiveDecorators.stories.ts [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+  decorators: [
+    (storyFn, context) => {
+      return () => h('div', [h('h2', ['Decorator not using args']), [h(storyFn())]]);
+    },
+  ],
+};
```
Comment (chakAs3) [inline code/renderers/vue3/template/stories/ReactiveDecorators.stories.ts:110]: @kasperpeulen Here you want to test a functional component but you use the Input to check if state is preseved ?
Normally Functional component is pure function so there is no state or lifecycle it just props -> vnode so 100% your input loses its value 
However the component is reactive to changes and not remounted

File: code/renderers/vue3/template/stories/ReactiveDecorators.stories.ts [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+export const DecoratorVNode: Story = {
+  args: { label: 'Decorator not using args' },
+  decorators: [
+    (storyFn, context) => {
+      return h('div', h(storyFn(context.args)));
+    },
+  ],
+};
+
+export const DecoratorVNodeArgsFromContext: Story = {
```
Comment (chakAs3) [inline code/renderers/vue3/template/stories/ReactiveDecorators.stories.ts:48]: Deecorators should be a Function that returns vnode or a Component.
Using VNode for component that wraps storyFn, you story will not be reactive. however the decorator is reactive in top usi level
The secret for reactivity is to be Wrapped in Function.

File: code/renderers/vue3/template/stories/ReactiveDecorators.stories.ts [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+          story: storyFn(context.args),
+        },
+        data() {
+          return { args: context.args };
+        },
+        template: '<div><h2>Decorator using label: {{args.label}}</h2><story/></div>',
+      });
+    },
+  ],
+};
```
Comment (chakAs3) [inline code/renderers/vue3/template/stories/ReactiveDecorators.stories.ts:88]: @kasperpeulen Here you want to test a functional component but you use the Input to check if state is preseved ? Normally Functional component is pure function so there is no state or lifecycle it just props -> vnode so 100% your input loses its value However the component is reactive to changes and not remountedDeecorators should be a Function that returns vnode or a Component.
Using VNode for component that wraps storyFn, you story will not be reactive. however the decorator is reactive in top usi
level
The secret for reactivity is to be Wrapped in Function.

File: code/renderers/vue3/template/stories/ReactiveDecorators.stories.ts [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+      return h({
+        components: {
+          story: storyFn(),
+        },
+        props: ['label'],
+        template: '<div><h1>Decorator using label: {{label}}</h1><story/></div>',
+      });
+    },
+  ],
+};
```
Comment (chakAs3) [inline code/renderers/vue3/template/stories/ReactiveDecorators.stories.ts:102]: Decorators should be either a Function that returns vnode or a Component.

* Using VNode for component that wraps storyFn, you story will not be reactive. however the decorator is reactive in top level
* NB: The secret for reactivity is to be Wrapped in Function.

File: code/renderers/vue3/template/stories/decorators.stories.js [hunk 7]
```diff
@@ -55,12 +60,19 @@ export const DynamicWrapper = {
...
-      data() {
-        return { level: `${args.level}px` };
-      },
-    }),
+    ComponentTemplateWrapper,
+    SimpleTemplateWrapper,
+    VueWrapperWrapper,
+    DynamicWrapperWrapper,
   ],
 };
```
Comment (chakAs3) [inline code/renderers/vue3/template/stories/decorators.stories.js:78]: @kasperpeulen this stories cover almost all  cases, i have done some changes and add story that stacks all of them which works perfect

**Review by chakAs3 — COMMENTED** [review]
@shilman 

Could you please review this PR as soon as possible? It primarily focuses on fixing reactivity in stories that use different decorators, functional components, templates, and other components.

I have added sufficient tests to cover various cases, including feedback from Kasper and modifications to ensure expected results.

I have a lot of work depending on this, and several users are requesting the merge after testing it.

Thank you for your time and assistance!

**Comment by thomasaull** [comment]
Thanks for your work on this @chakAs3!

File: code/renderers/vue3/template/stories/ReactiveDecorators.stories.ts [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+
+/**
+ * Decorator should either be a function that returns a VNode or a component
+ * so using VNode is not recommended to wrap the story if you want to use args in reactive mode
+ */
+export const DecoratorVNode: Story = {
+  args: { label: 'Decorator not using args' },
+  decorators: [
+    (storyFn, context) => {
+      return h('div', h(storyFn(context.args)));
```
Comment (kasperpeulen) [inline code/renderers/vue3/template/stories/ReactiveDecorators.stories.ts:43]: I think I originally wrote this test without passing in context.args. 
I think passing in context.args should not be required right?

Also `storyFn` should a function accepting context and not args right? cc @tmeasday

Comment (chakAs3) [reply code/renderers/vue3/template/stories/ReactiveDecorators.stories.ts]: yes i removed it, i was just for some testing purpose

**Review by kasperpeulen — APPROVED** [review]
If you can get the tests in proper state, feel free to merge it.

File: code/renderers/vue3/src/render.ts [hunk 9]
```diff
@@ -22,65 +23,84 @@ export const setup = (fn: (app: any) => void) => {
...
-  const storybookApp = createApp({
-    render() {
-      const renderedElement: any = elementMap.get(canvasElement);
-      const current = renderedElement && renderedElement.template ? renderedElement : element;
-      map.set(canvasElement, { vueApp: storybookApp, reactiveArgs });
-      return h(current, reactiveArgs);
+  // create vue app for the story
+  const vueApp = createApp({
+    setup() {
+      storyContext.args = reactive(storyContext.args);
```
[the remark this answers is not in the document]
Comment (chakAs3) [reply code/renderers/vue3/src/render.ts:56]: > I just tried it out, but it seems like you can not use globals in a reactive way. For example:
> 
> ```ts
>   render: (args, context) => {
>     return h('div', [
>       'Custom render is a functional component',
>       JSON.stringify(context.globals),
>       h(Reactivity, args),
>     ]);
>   },
> ```

i tried to watch context.globals  changes but it is not necessary as core render i intercept  globals changes anyway in renderToCanvas() and i trigger the unboundStoryFn in order for react components to rerender with new globals value.
if a user want to use context.globals in his story or decorator  his has to handle the reactivity him the way he wants, ref , reactive,.. or any primitive 
example user in Vuetify implementation :
```ts
// This file is a decorator that wraps the story in a VuetifyAppWrapper
import { h , shallowReactive} from 'vue';
import StoryWrapper from './VuetifyAppWrapper.vue';

export const DEFAULT_THEME = 'light';

const theme = shallowReactive({themeName:DEFAULT_THEME});

export const withVuetifyTheme = (storyFn, context) => {
  // Pull our global theme variable, fallback to DEFAULT_THEME
  theme.themeName =  context.globals.theme || DEFAULT_THEME;
  const story = storyFn();

  return () => {
    return h(
      StoryWrapper,
      // give themeName to StoryWrapper as a prop
      theme,
      {
        story: () => h(story, { ...context.args }),
      }
    );
  };
};

```

**Comment by chakAs3** [comment]
> If you can get the tests in proper state, feel free to merge it.

i have gone all the tests to get them properly, + ReactiveSlots.stories as well

**Pull request merged**